added download, edit and retry buttons to upload modal

gitolicious opened the pull request:
## Description:
After a failed upload, the following options are usually taken:
- Download the bin file and upload it manually (OTA not possible, network issues, ...)
- Edit the config yaml
- Retry upload
I added buttons for these options to the upload modal. On thing to note is that, when checking if the upload succeeded, only compilation is taken into consideration at the moment, not the actual upload. I think this could be improved, then also button display can be finetuned to the error that actually occurred.

**Related issue (if applicable):** n/a

**Pull request in [esphome-docs](https://github.com/esphome/esphome-docs) with documentation (if applicable):** n/a

## Checklist:
  - [x] The code change is tested and works locally.
  - [n/a] Tests have been added to verify that the new code works (under `tests/` folder).

If user exposed functionality or configuration variables are added/changed:
  - [n/a] Documentation added/updated in [esphome-docs](https://github.com/esphome/esphome-docs).

Commit:
added download, edit and retry buttons to upload modal

OttoWinter commented on esphome/dashboard/static/esphome.js:
```diff
@@ -406,16 +406,27 @@ const logsModal = new LogModalElem({
   name: 'upload',
   onPrepare: (modalElem, config) => {
+    retryUploadButton.setAttribute('data-node', uploadModal.activeConfig);
+    retryUploadButton.classList.add('hide');
```
Let's not hide anything - otherwise the buttons will reposition themselves, which is a big red flag.

gitolicious replied:
Ok, I will change to disabled

gitolicious replied:
Done.

OttoWinter commented on esphome/dashboard/static/esphome.js:
```diff
@@ -406,16 +406,27 @@ const logsModal = new LogModalElem({
   onProcessExit: (modalElem, code) => {
     if (code === 0) {
       M.toast({html: "Program exited successfully."});
+      downloadAfterUploadButton.classList.remove('disabled');
```
Why have a download button if OTA worked *successfully*? I mean the use-case here would be if OTA fails? I'm not so sure about this button either. Having 4 buttons at the bottom is a bit too much.

gitolicious replied:
Check my description. "when checking if the upload succeeded, only compilation is taken into consideration at the moment, not the actual upload"
If you can help me fix this, the button should only be shown when OTA failed.

Background: Compilation succeeds, but OTA fails (I have this frequently for various reasons: DNS issues, device renaming, ...) you will need to close the upload window, find the entry in the dashboard list, click on the menu button, click on compile, wait for it to finish which is recompiling the exact same thing as you had before, then download.

I would also be ok if there was any other option to download the latest compiled bin easily.

About the button number: If you think it's too much, we could put some of them in an expandable menu at the bottom right of the compilation modal?

OttoWinter replied:
Ah ok missed that.

Yeah that's not going to be possible to check easily - the process is spawned and all communication happens via stdout. One possibility would of course be to check for some text in stdout, but I don't think that would be worth the added complexity.

Also from another point I think it makes sense to hold it in COMPILE: All guides reference COMPILE, changing it would only confuse users. Plus from a quick check of this PR in my local environment 4 buttons are definitely too much.

Commit:
changed hide for disabled, resorted buttons

Commit:
moved upload action buttons to menu

gitolicious commented:
I created a menu in the upload modal and put the buttons there. What do you think?
![image](https://user-images.githubusercontent.com/26963495/58437496-8420c500-80ca-11e9-9028-7c58dd9f1c99.png)

![image](https://user-images.githubusercontent.com/26963495/58437515-93a00e00-80ca-11e9-8ddc-ca7e12ef93c9.png)

OttoWinter commented:
I'm not so sure about the placement of the overflow button - if i remember the material design guide correctly the CLOSE/END button should always be in the bottom right. But... we're not following md spec that closely anyway. If we find a better place for it, we can still move it.

For the buttons in there: I think the "edit" button is useful enough that it can be displayed next to "CLOSE" (so in the main bar). At least in the last few days I've really learned to love that EDIT button (thanks btw for that!).

While we're at it: One thing we could also do is change the SAVE&VALIDATE button in the editor to a SAVE&UPLOAD button. Reason is: the editor now performs validation on-the-fly, if you look at the left gutter you should see live validation. If you want, you can put it in this PR, though I don't want to force you to more work :)

Commit:
button changes

moved edit button from menu in upload modal
swapped validation button for upload in edit modal

gitolicious commented:
Feel free to reposition the menu if you find a better place. For now I can't think of one...

Edit button has been moved back from menu to the main buttons.

About validation: I just saw the new live validation feature yesterday - it's awesome! Does it need to be called _SAVE & UPLOAD_ or would _UPLOAD_ be sufficient? What do you think about my idea of disabling it when live validation is failing?

And don't worry - that was an easy fix I was know my way around that part of the code now.

OttoWinter commented:
> Does it need to be called SAVE & UPLOAD or would UPLOAD be sufficient?

Upload is also good, yeah.

> What do you think about my idea of disabling it when live validation is failing?

Hmm yeah that could be possible. There's one thing I would be a bit worried about though: What does it look like while typing? If the button constantly flashes disabled/enabled while typing that could be bad. But I also haven't tried so it might be ok.

I think I'll merge this PR as-is now because it looks like a neat feature and would fit well in the next beta - if you want to change other things feel free to open a new PR.

gitolicious commented:
Ok, thx for merging. I love that I can use the new features in my Hass.io ESPHome dev add-ons only minutes later!
